File: python/tcp/servidor_tcp.py
# This is a sample Python script.
import socket
from datetime import datetime

HOST = "0.0.0.0"
PORTA = 5000

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as servidor:
    servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    servidor.bind((HOST, PORTA))
    servidor.listen(1)

    print(f"[TCP] Servidor aguardando conexões na porta {PORTA}...")

    conexao, endereco = servidor.accept()

    with conexao:
        print(f"[TCP] Cliente conectado: {endereco}")

        while True:
            dados = conexao.recv(1024).decode("utf-8").strip()

            if not dados:
                break

            print(f"[TCP] Recebido: {dados}")

            if dados.lower() == "sair":
                conexao.sendall(
                    "Encerrando conexão. Até mais!\n".encode("utf-8")
                )
                break

            elif dados.lower() == "hora":
                horario_atual = datetime.now().strftime("%H:%M:%S")

                resposta = (
                    f"Horário atual do servidor: {horario_atual}\n"
                )

            else:
                resposta = (
                    f'Monitor responde: recebi sua mensagem -> "{dados}"\n'
                )

            conexao.sendall(resposta.encode("utf-8"))

print("[TCP] Servidor encerrado.")

File: java/tcp/ServidorTCP.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;

import java.io.*;
import java.net.*;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;

public class ServidorTCP {

    public static void main(String[] args) throws IOException {

        int porta = 5000;

        try (ServerSocket servidor = new ServerSocket(porta)) {

            System.out.println("[TCP] Servidor aguardando conexões na porta " + porta + "...");

            try (Socket cliente = servidor.accept();
                 BufferedReader entrada = new BufferedReader(
                         new InputStreamReader(cliente.getInputStream()));
                 PrintWriter saida = new PrintWriter(
                         cliente.getOutputStream(), true)) {

                System.out.println("[TCP] Cliente conectado: "
                        + cliente.getRemoteSocketAddress());

                String mensagem;

                while ((mensagem = entrada.readLine()) != null) {

                    System.out.println("[TCP] Recebido: " + mensagem);

                    // Encerra a conexão
                    if (mensagem.equalsIgnoreCase("sair")) {
                        saida.println("Encerrando conexão. Até mais!");
                        break;
                    }

                    // Retorna o horário atual do servidor
                    if (mensagem.equalsIgnoreCase("hora")) {

                        LocalTime horarioAtual = LocalTime.now();

                        DateTimeFormatter formato =
                                DateTimeFormatter.ofPattern("HH:mm:ss");

                        saida.println(
                                "Horário atual do servidor: "
                                        + horarioAtual.format(formato)
                        );

                    } else {

                        // Resposta padrão
                        saida.println(
                                "Monitor responde: recebi sua mensagem -> \""
                                        + mensagem + "\""
                        );
                    }
                }
            }
        }

        System.out.println("[TCP] Servidor encerrado.");
    }
}